### ISSUE TITLE
Various raft tests fail occasionally

### ISSUE THREAD

**Issue by abligh** [hashicorp/raft#90]
Background: I am testing a new transport and to get this to work need it to pass all tests. Unfortunately some tests fail occasionally with the existing InmemoryTransport.

Running
```
for i in `seq 1 50` ; do go test -run TestRaft_Voting ; done
```

I see the occasional failure, like:

```
--- FAIL: TestRaft_Voting (0.09s)
	raft_test.go:189: [WARN] Fully Connecting
	raft_test.go:93: [INFO] raft: Node at efecfb3b-81c1-a503-46c3-05fbf14221b6 [Follower] entering Follower state
	raft_test.go:93: [INFO] raft: Node at e6ba3ba4-a326-f7d0-e260-9d1cc55719d0 [Follower] entering Follower state
	raft_test.go:93: [INFO] raft: Node at 2696da81-b3ca-69e1-b09e-ef3f91fed6ea [Follower] entering Follower state
	raft_test.go:93: [WARN] raft: Heartbeat timeout reached, starting election
	raft_test.go:93: [INFO] raft: Node at e6ba3ba4-a326-f7d0-e260-9d1cc55719d0 [Candidate] entering Candidate state
	raft_test.go:93: [DEBUG] raft: Votes needed: 2
	raft_test.go:93: [DEBUG] raft: Vote granted from e6ba3ba4-a326-f7d0-e260-9d1cc55719d0. Tally: 1
	raft_test.go:93: [DEBUG] raft: Vote granted from efecfb3b-81c1-a503-46c3-05fbf14221b6. Tally: 2
	raft_test.go:93: [INFO] raft: Election won. Tally: 2
	raft_test.go:93: [INFO] raft: Node at e6ba3ba4-a326-f7d0-e260-9d1cc55719d0 [Leader] entering Leader state
	raft_test.go:93: [WARN] raft: Heartbeat timeout reached, starting election
	raft_test.go:93: [INFO] raft: Node at 2696da81-b3ca-69e1-b09e-ef3f91fed6ea [Candidate] entering Candidate state
	raft_test.go:93: [DEBUG] raft: Votes needed: 2
	raft_test.go:93: [DEBUG] raft: Vote granted from 2696da81-b3ca-69e1-b09e-ef3f91fed6ea. Tally: 1
	raft_test.go:93: [ERR] raft: peer 2696da81-b3ca-69e1-b09e-ef3f91fed6ea has newer term, stopping replication
	raft_test.go:93: [INFO] raft: pipelining replication to peer efecfb3b-81c1-a503-46c3-05fbf14221b6
	raft_test.go:93: [DEBUG] raft: Node e6ba3ba4-a326-f7d0-e260-9d1cc55719d0 updated peer set (2): [e6ba3ba4-a326-f7d0-e260-9d1cc55719d0 efecfb3b-81c1-a503-46c3-05fbf14221b6 2696da81-b3ca-69e1-b09e-ef3f91fed6ea]
	raft_test.go:93: [WARN] raft: Rejecting vote request from 2696da81-b3ca-69e1-b09e-ef3f91fed6ea since we have a leader: e6ba3ba4-a326-f7d0-e260-9d1cc55719d0
	raft_test.go:93: [INFO] raft: Node at e6ba3ba4-a326-f7d0-e260-9d1cc55719d0 [Follower] entering Follower state
	raft_test.go:93: [INFO] raft: Node at 2696da81-b3ca-69e1-b09e-ef3f91fed6ea [Follower] entering Follower state
	raft_test.go:1661: expected vote not to be granted, but was {Term:42 Peers:[146 218 0 36 101 102 101 99 102 98 51 98 45 56 49 99 49 45 97 53 48 51 45 52 54 99 51 45 48 53 102 98 102 49 52 50 50 49 98 54 218 0 36 101 54 98 97 51 98 97 52 45 97 51 50 54 45 102 55 100 48 45 101 50 54 48 45 57 100 49 99 99 53 53 55 49 57 100 48] Granted:true}
	raft_test.go:93: [INFO] raft: aborting pipeline replication to peer efecfb3b-81c1-a503-46c3-05fbf14221b6
FAIL
exit status 1
FAIL	github.com/hashicorp/raft	0.101s
```

**Comment by abligh** [comment]
Here's an even simpler example, this time from `TestRaft_Barrier`:

To make this fail reliably, I compiled the test first, then ran the compiled version:
```
go test -c -v
for i in `seq 1 1000`; do ./raft.test -test.v=1 -test.run='TestRaft_Barrier' >> z2.inmem 2>&1 ; done
```

Looking for `FAIL:` in the output file (here `z2.inmem`) I see a lot of:

```
--- FAIL: TestRaft_Barrier (0.07s)
        raft_test.go:192: [WARN] Fully Connecting
        raft_test.go:96: [INFO] raft: Node at a291f41a-d0da-311f-8fb0-c5ea96036173 [Follower] entering Follower state
        raft_test.go:96: [INFO] raft: Node at 8f73c435-5045-2b0d-5225-2c5a8ff8d628 [Follower] entering Follower state
        raft_test.go:96: [INFO] raft: Node at 9fd38b33-2aea-2bf0-699c-ffaa61291aa5 [Follower] entering Follower state
        raft_test.go:96: [WARN] raft: Heartbeat timeout reached, starting election
        raft_test.go:96: [INFO] raft: Node at 8f73c435-5045-2b0d-5225-2c5a8ff8d628 [Candidate] entering Candidate state
        raft_test.go:96: [DEBUG] raft: Votes needed: 2
        raft_test.go:96: [DEBUG] raft: Vote granted from 8f73c435-5045-2b0d-5225-2c5a8ff8d628. Tally: 1
        raft_test.go:96: [DEBUG] raft: Vote granted from a291f41a-d0da-311f-8fb0-c5ea96036173. Tally: 2
        raft_test.go:96: [INFO] raft: Election won. Tally: 2
        raft_test.go:96: [INFO] raft: Node at 8f73c435-5045-2b0d-5225-2c5a8ff8d628 [Leader] entering Leader state
        raft_test.go:96: [WARN] raft: Heartbeat timeout reached, starting election
        raft_test.go:96: [INFO] raft: Node at a291f41a-d0da-311f-8fb0-c5ea96036173 [Candidate] entering Candidate state
        raft_test.go:96: [DEBUG] raft: Votes needed: 2
        raft_test.go:96: [DEBUG] raft: Vote granted from a291f41a-d0da-311f-8fb0-c5ea96036173. Tally: 1
        raft_test.go:96: [WARN] raft: Rejecting vote request from a291f41a-d0da-311f-8fb0-c5ea96036173 since we have a leader: 8f73c435-5045-2b0d-5225-2c5a8ff8d628
        raft_test.go:96: [WARN] raft: Rejecting vote request from a291f41a-d0da-311f-8fb0-c5ea96036173 since we have a leader: 8f73c435-5045-2b0d-5225-2c5a8ff8d628
        raft_test.go:96: [ERR] raft: peer a291f41a-d0da-311f-8fb0-c5ea96036173 has newer term, stopping replication
        raft_test.go:96: [INFO] raft: pipelining replication to peer 9fd38b33-2aea-2bf0-699c-ffaa61291aa5
        raft_test.go:96: [DEBUG] raft: Node 8f73c435-5045-2b0d-5225-2c5a8ff8d628 updated peer set (2): [8f73c435-5045-2b0d-5225-2c5a8ff8d628 a291f41a-d0da-311f-8fb0-c5ea96036173 9fd38b33-2aea-2bf0-699c-ffaa61291aa5]
        raft_test.go:96: [INFO] raft: Node at 8f73c435-5045-2b0d-5225-2c5a8ff8d628 [Follower] entering Follower state
        raft_test.go:96: [INFO] raft: aborting pipeline replication to peer 9fd38b33-2aea-2bf0-699c-ffaa61291aa5
        raft_test.go:1560: err: node is not the leader
```

And similar.

I think what is happening is (roughly) as follows:
* Node `8f73c435-5045-2b0d-5225-2c5a8ff8d628` goes into candidate state and starts an election
* Node `8f73c435-5045-2b0d-5225-2c5a8ff8d628` votes for itself
* Node `a291f41a-d0da-311f-8fb0-c5ea96036173` votes for `8f73c435-5045-2b0d-5225-2c5a8ff8d628`
* Node `8f73c435-5045-2b0d-5225-2c5a8ff8d628` considers itself leader, having two votes

So far so good. Here's where things go wrong.
* Node `a291f41a-d0da-311f-8fb0-c5ea96036173` says its heartbeat timeout has occurred. It shouldn't do, because it's just voted for the current leader.

**Comment by abligh** [comment]
Here's another one with microsecond time logs (logging changed to go straight to stderr too so appears differently in sequence to test output):

```
=== RUN TestRaft_Barrier
14:03:00.104428 [INFO] raft: Node at 5547e43d-9f14-5e15-9eb9-b9cac4448688 [Follower] entering Follower state
14:03:00.104526 [INFO] raft: Node at bc57f297-48fc-8ccc-d122-f4b1aea9fe46 [Follower] entering Follower state
14:03:00.104563 [INFO] raft: Node at 2f262c87-1bdb-99e4-403a-de8767c41ae0 [Follower] entering Follower state
14:03:00.177835 [WARN] raft: Heartbeat timeout reached, starting election
14:03:00.177888 [INFO] raft: Node at bc57f297-48fc-8ccc-d122-f4b1aea9fe46 [Candidate] entering Candidate state
14:03:00.177947 [DEBUG] raft: Votes needed: 2
14:03:00.177972 [DEBUG] raft: Vote granted from bc57f297-48fc-8ccc-d122-f4b1aea9fe46. Tally: 1
14:03:00.178211 [DEBUG] raft: Vote granted from 5547e43d-9f14-5e15-9eb9-b9cac4448688. Tally: 2
14:03:00.178225 [INFO] raft: Election won. Tally: 2
14:03:00.178235 [INFO] raft: Node at bc57f297-48fc-8ccc-d122-f4b1aea9fe46 [Leader] entering Leader state
14:03:00.179009 [WARN] raft: Heartbeat timeout reached, starting election
14:03:00.179058 [INFO] raft: Node at 5547e43d-9f14-5e15-9eb9-b9cac4448688 [Candidate] entering Candidate state
14:03:00.179076 [DEBUG] raft: Votes needed: 2
14:03:00.179084 [DEBUG] raft: Vote granted from 5547e43d-9f14-5e15-9eb9-b9cac4448688. Tally: 1
14:03:00.179148 [ERR] raft: peer 5547e43d-9f14-5e15-9eb9-b9cac4448688 has newer term, stopping replication
14:03:00.179167 [INFO] raft: pipelining replication to peer 2f262c87-1bdb-99e4-403a-de8767c41ae0
14:03:00.179250 [WARN] raft: Rejecting vote request from 5547e43d-9f14-5e15-9eb9-b9cac4448688 since we have a leader: bc57f297-48fc-8ccc-d122-
f4b1aea9fe46
14:03:00.179310 [DEBUG] raft: Node bc57f297-48fc-8ccc-d122-f4b1aea9fe46 updated peer set (2): [bc57f297-48fc-8ccc-d122-f4b1aea9fe46 5547e43d-9
f14-5e15-9eb9-b9cac4448688 2f262c87-1bdb-99e4-403a-de8767c41ae0]
14:03:00.179351 [INFO] raft: Node at bc57f297-48fc-8ccc-d122-f4b1aea9fe46 [Follower] entering Follower state
14:03:00.179390 [WARN] raft: Rejecting vote request from 5547e43d-9f14-5e15-9eb9-b9cac4448688 since we have a leader: bc57f297-48fc-8ccc-d122-
f4b1aea9fe46
14:03:00.179438 [INFO] raft: aborting pipeline replication to peer 2f262c87-1bdb-99e4-403a-de8767c41ae0
--- FAIL: TestRaft_Barrier (0.08s)
        raft_test.go:193: [WARN] Fully Connecting
        raft_test.go:1561: err: node is not the leader
```

Node `5547e43d-9f14-5e15-9eb9-b9cac4448688` decides for some reason to start an election only a millisecond after voting.

**Comment by abligh** [comment]
Pull request #91 fixes this.

**Comment by armon** [comment]
Fixed by #91. Thanks!
